File: compiler/bench/amicable.c
// expected: 31626
// Amicable pairs benchmark - sum of amicable pairs below 10000, run 5 times

#include <stdio.h>

int proper_divisor_sum(int n) {
    if (n == 1) return 0;
    int sum = 1;
    for (int i = 2; i <= n / 2; i++) {
        if (n % i == 0) sum += i;
    }
    return sum;
}

long amicable_sum(int limit) {
    long sum = 0;
    for (int i = 2; i < limit; i++) {
        int s = proper_divisor_sum(i);
        if (s > i && proper_divisor_sum(s) == i) {
            sum += i + s;
        }
    }
    return sum;
}

int main(void) {
    long result = 0;
    for (int i = 0; i < 5; i++) {
        result = amicable_sum(10000);
    }
    printf("%ld\n", result);
    return 0;
}

File: compiler/bench/amicable.fs
\ expected: 31626
\ Amicable pairs benchmark - sum of amicable pairs below 10000, run 5 times

: proper-divisor-sum ( n -- sum )
  dup 1 = if drop 0 exit then
  1 swap
  dup 2/ 2 do
    dup i mod 0= if i under+ then
  loop drop ;

: amicable-sum ( limit -- sum )
  0 swap 2 do
    i proper-divisor-sum
    dup i > if
      dup proper-divisor-sum i = if
        dup i + under+
      else drop then
    else drop then
  loop ;

: main ( -- )
  0
  5 0 do
    drop 10000 amicable-sum
  loop
  . cr ;
